File: web/apps/viewer/src/lib/browser-artifacts.ts
export type BrowserPoint = {
  x: number;
  y: number;
};

export type BrowserRect = BrowserPoint & {
  width: number;
  height: number;
};

export type BrowserProvenanceSpan =
  | {
      kind: "file";
      path: string;
      start_utf8: number;
      end_utf8: number;
    }
  | {
      kind: "generated";
      stable_id: string;
      description: string;
    };

export type BrowserSourceProvenance = {
  primary: BrowserProvenanceSpan;
  related: unknown[];
  expansion_stack: unknown[];
  generated_by: string;
  expansion_stack_truncated: boolean;
};

export type BrowserFontFamily =
  | "serif"
  | "sans"
  | "mono"
  | "math"
  | "symbol"
  | "math_extension"
  | { named: string };

export type BrowserFontRequest = {
  family: BrowserFontFamily;
  series: "regular" | "bold";
  shape: "upright" | "italic";
  size_pt: number;
  role: "body" | "heading" | "math" | "mono";
};

export type BrowserDrawOp =
  | { kind: "save" }
  | { kind: "restore" }
  | ({ kind: "clip_rect" } & BrowserRect)
  | {
      kind: "text_run";
      origin: BrowserPoint;
      text: string;
      font: BrowserFontRequest;
      size_pt: number;
      approximate_advance_pt: number;
      resolved_font?: {
        face_id: string;
        postscript_name: string;
        glyph_id_kind: "type1_char_code" | "open_type_glyph_id";
        content_hash: string;
      };
      glyphs?: Array<{
        glyph_id: number;
        advance_pt: number;
        offset: BrowserPoint;
      }>;
      clusters?: Array<{
        text_start_utf8: number;
        text_end_utf8: number;
        glyph_start: number;
        glyph_end: number;
      }>;
      source: BrowserSourceProvenance;
    }
  | ({ kind: "rule" } & BrowserRect)
  | {
      kind: "image";
      rect: BrowserRect;
      asset_ref: string;
      asset_format?: string;
      page_selection?: unknown;
      asset_hash?: string;
      natural_width_pt?: number;
      natural_height_pt?: number;
      crop?: unknown;
      scale?: { x: number; y: number };
      rotation?: {
        angle_degrees: number;
        origin?: string;
      };
      diagnostic?: string;
      source: BrowserSourceProvenance;
    }
  | {
      kind: "link_annotation";
      rect: BrowserRect;
      target: string;
      source: BrowserSourceProvenance;
    }
  | {
      kind: "named_destination";
      name: string;
      point: BrowserPoint;
      source: BrowserSourceProvenance;
    };

export type BrowserPageDisplayList = {
  page_id: string;
  width_pt: number;
  height_pt: number;
  ops: BrowserDrawOp[];
  source_spans: unknown[];
  content_hash: string;
};

export type BrowserAssetManifestEntry = {
  asset_ref: string;
  format?: string;
  content_hash?: string;
};

export type BrowserPagesArtifact = {
  schema_version: number;
  revision: number;
  pages: BrowserPageDisplayList[];
  changed_page_ids: string[];
  removed_page_ids: string[];
  assets: BrowserAssetManifestEntry[];
};

export type BrowserBuildMetadata = {
  schema_version: number;
  revision: number;
  compile_mode: "one_shot" | "incremental";
  event_count: number;
  diagnostic_count: number;
  pages: {
    total: number;
    changed: number;
    reused: number;
    removed: number;
  };
};

export type BrowserArtifactExpectations = {
  revision: number;
  page_count: number;
  event_count: number;
  diagnostic_count: number;
};

export function validateBrowserArtifacts(
  pageArtifact: BrowserPagesArtifact,
  buildMetadata: BrowserBuildMetadata,
  expected: BrowserArtifactExpectations
) {
  if (pageArtifact.schema_version !== 1 || buildMetadata.schema_version !== 1) {
    throw new Error("WASI compiler returned an unsupported browser artifact schema");
  }
  if (pageArtifact.revision !== expected.revision || buildMetadata.revision !== expected.revision) {
    throw new Error("WASI compiler returned browser artifacts for the wrong revision");
  }
  if (
    buildMetadata.compile_mode !== "one_shot"
    || expected.page_count !== pageArtifact.pages.length
    || buildMetadata.pages.total !== pageArtifact.pages.length
    || buildMetadata.pages.changed !== pageArtifact.changed_page_ids.length
    || buildMetadata.pages.reused !== 0
    || buildMetadata.pages.removed !== pageArtifact.removed_page_ids.length
    || buildMetadata.event_count !== expected.event_count
    || buildMetadata.diagnostic_count !== expected.diagnostic_count
  ) {
    throw new Error("WASI compiler returned inconsistent browser page counts");
  }
  const pageIds = new Set(pageArtifact.pages.map((page) => page.page_id));
  const changedPageIds = new Set(pageArtifact.changed_page_ids);
  const removedPageIds = new Set(pageArtifact.removed_page_ids);
  if (
    pageIds.size !== pageArtifact.pages.length
    || changedPageIds.size !== pageArtifact.changed_page_ids.length
    || removedPageIds.size !== pageArtifact.removed_page_ids.length
    || changedPageIds.size !== pageIds.size
    || pageArtifact.pages.some((page) => (
      page.page_id.length === 0
      || page.content_hash.length === 0
      || !Number.isFinite(page.width_pt)
      || page.width_pt <= 0
      || !Number.isFinite(page.height_pt)
      || page.height_pt <= 0
    ))
    || pageArtifact.changed_page_ids.some((pageId) => !pageIds.has(pageId))
    || pageArtifact.removed_page_ids.some((pageId) => pageIds.has(pageId))
  ) {
    throw new Error("WASI compiler returned an invalid browser page manifest");
  }

  const assets: unknown = pageArtifact.assets;
  if (!Array.isArray(assets)) {
    throw new Error("WASI compiler returned an invalid browser asset manifest");
  }
  const assetRefs = new Set<string>();
  for (const asset of assets) {
    if (!asset || typeof asset !== "object") {
      throw new Error("WASI compiler returned an invalid browser asset manifest");
    }
    const entry = asset as Record<string, unknown>;
    if (
      typeof entry.asset_ref !== "string"
      || entry.asset_ref.length === 0
      || (entry.format !== undefined && (
        typeof entry.format !== "string"
        || entry.format.length === 0
      ))
      || (entry.content_hash !== undefined && (
        typeof entry.content_hash !== "string"
        || entry.content_hash.length === 0
      ))
      || assetRefs.has(entry.asset_ref)
    ) {
      throw new Error("WASI compiler returned an invalid browser asset manifest");
    }
    assetRefs.add(entry.asset_ref);
  }
}

File: crates/tex-render-model/tests/browser_artifacts.rs
use tex_render_model::{
    BrowserBuildMetadata, BrowserCompileMode, BrowserFontAsset, BrowserGlyphOutline,
    BrowserPagesArtifact, DrawOp, FontFaceId, GlyphIdKind, GlyphOutlineCommand, GraphicAssetFormat,
    PageDisplayList, PositionedImage, Rect, SourceProvenance,
};

fn page(page_id: &str, content_hash: &str, ops: Vec<DrawOp>) -> PageDisplayList {
    PageDisplayList {
        page_id: page_id.to_string(),
        width_pt: 612.0,
        height_pt: 792.0,
        ops,
        source_spans: Vec::new(),
        content_hash: content_hash.to_string(),
    }
}

#[test]
fn one_shot_browser_artifacts_preserve_compiler_page_identity_and_assets() {
    let pages = vec![
        page("page-a", "hash-a", Vec::new()),
        page(
            "page-b",
            "hash-b",
            vec![DrawOp::Image(PositionedImage {
                rect: Rect {
                    x: 36.0,
                    y: 72.0,
                    width: 120.0,
                    height: 90.0,
                },
                asset_ref: "figures/result.png".to_string(),
                asset_format: Some(GraphicAssetFormat::Png),
                page_selection: None,
                asset_hash: Some("asset-hash".to_string()),
                natural_width_pt: Some(240.0),
                natural_height_pt: Some(180.0),
                crop: None,
                scale: None,
                rotation: None,
                diagnostic: None,
                source: SourceProvenance::generated("figure", "test figure"),
            })],
        ),
    ];

    let font = BrowserFontAsset {
        face_id: FontFaceId::new("cmr10"),
        postscript_name: "CMR10".to_string(),
        glyph_id_kind: GlyphIdKind::Type1CharCode,
        content_hash: "blake3:font".to_string(),
        glyphs: vec![BrowserGlyphOutline {
            glyph_id: u32::from(b'A'),
            commands: vec![
                GlyphOutlineCommand::MoveTo { x: 0.0, y: 0.0 },
                GlyphOutlineCommand::LineTo { x: 0.5, y: 1.0 },
                GlyphOutlineCommand::Close,
            ],
        }],
    };
    let artifact = BrowserPagesArtifact::one_shot(17, pages, vec![font]);

    assert_eq!(artifact.schema_version, 2);
    assert_eq!(artifact.revision, 17);
    assert_eq!(artifact.changed_page_ids, ["page-a", "page-b"]);
    assert!(artifact.removed_page_ids.is_empty());
    assert_eq!(artifact.pages[1].page_id, "page-b");
    assert_eq!(artifact.pages[1].content_hash, "hash-b");
    assert_eq!(artifact.assets.len(), 1);
    assert_eq!(artifact.assets[0].asset_ref, "figures/result.png");
    assert_eq!(artifact.assets[0].format, Some(GraphicAssetFormat::Png));
    assert_eq!(
        artifact.assets[0].content_hash.as_deref(),
        Some("asset-hash")
    );
    assert_eq!(artifact.fonts.len(), 1);
    assert_eq!(artifact.fonts[0].face_id.as_str(), "cmr10");
    assert_eq!(artifact.fonts[0].glyphs[0].glyph_id, u32::from(b'A'));

    let metadata = BrowserBuildMetadata::one_shot(17, 42, 3, &artifact);
    assert_eq!(metadata.schema_version, 1);
    assert_eq!(metadata.revision, 17);
    assert_eq!(metadata.compile_mode, BrowserCompileMode::OneShot);
    assert_eq!(metadata.event_count, 42);
    assert_eq!(metadata.diagnostic_count, 3);
    assert_eq!(metadata.pages.total, 2);
    assert_eq!(metadata.pages.changed, 2);
    assert_eq!(metadata.pages.reused, 0);
    assert_eq!(metadata.pages.removed, 0);

    let encoded = serde_json::to_string(&artifact).expect("serialize browser page artifact");
    let decoded: BrowserPagesArtifact =
        serde_json::from_str(&encoded).expect("deserialize browser page artifact");
    assert_eq!(decoded, artifact);
}

#[test]
fn browser_asset_manifest_deduplicates_repeated_image_references() {
    let image = DrawOp::Image(PositionedImage {
        rect: Rect {
            x: 0.0,
            y: 0.0,
            width: 10.0,
            height: 10.0,
        },
        asset_ref: "figure.pdf".to_string(),
        asset_format: Some(GraphicAssetFormat::Pdf),
        page_selection: None,
        asset_hash: Some("pdf-hash".to_string()),
        natural_width_pt: None,
        natural_height_pt: None,
        crop: None,
        scale: None,
        rotation: None,
        diagnostic: None,
        source: SourceProvenance::generated("figure", "test figure"),
    });
    let artifact = BrowserPagesArtifact::one_shot(
        1,
        vec![
            page("page-a", "hash-a", vec![image.clone()]),
            page("page-b", "hash-b", vec![image]),
        ],
        Vec::new(),
    );

    assert_eq!(artifact.assets.len(), 1);
    assert_eq!(artifact.assets[0].asset_ref, "figure.pdf");
    assert_eq!(artifact.assets[0].format, Some(GraphicAssetFormat::Pdf));
    assert_eq!(artifact.assets[0].content_hash.as_deref(), Some("pdf-hash"));
}
